File: remix/src/core/url.ts
/**
 * URL helpers for redirects.
 */

const DEFAULT_VIDEO_URL = 'https://showad.proofmark.io';

export function resolveVideoAdUrl(videoAdUrl?: string): string {
  return (
    videoAdUrl ||
    (typeof process !== 'undefined' && process.env?.SHOWAD_VIDEO_URL) ||
    DEFAULT_VIDEO_URL
  );
}

/**
 * Build the URL the visitor is sent to in order to watch the gating video ad.
 * Format: `${videoAdUrl}/c/${creatorHash}?return_url=...&sdk=1`.
 */
export function buildVideoAdRedirectUrl(input: {
  videoAdUrl?: string;
  creatorHash: string;
  returnUrl: string;
}): string {
  const url = new URL(`/c/${input.creatorHash}`, resolveVideoAdUrl(input.videoAdUrl));
  url.searchParams.set('return_url', input.returnUrl);
  url.searchParams.set('sdk', '1');
  return url.toString();
}

/** Resource-scoped variant: `/c/<creator>/<project>/<resource>`. */
export function buildResourceRedirectUrl(input: {
  videoAdUrl?: string;
  creatorHash: string;
  projectHash: string;
  resourceHash: string;
  returnUrl: string;
}): string {
  const url = new URL(
    `/c/${input.creatorHash}/${input.projectHash}/${input.resourceHash}`,
    resolveVideoAdUrl(input.videoAdUrl)
  );
  url.searchParams.set('return_url', input.returnUrl);
  url.searchParams.set('sdk', '1');
  return url.toString();
}

/** Return a copy of `url` with the given query param removed. */
export function removeQueryParam(url: string, paramName: string): string {
  try {
    const parsed = new URL(url);
    parsed.searchParams.delete(paramName);
    return parsed.toString();
  } catch {
    return url;
  }
}

File: python/src/proofmark_showad/url.py
"""URL helpers."""

from __future__ import annotations

from typing import Optional
from urllib.parse import quote, urlencode, urlparse, urlunparse, parse_qsl


def _trim_slash(url: str) -> str:
    return url.rstrip("/")


def build_video_ad_redirect_url(
    video_ad_url: str,
    creator_hash: str,
    return_url: Optional[str],
) -> str:
    base = _trim_slash(video_ad_url) + "/c/" + quote(creator_hash, safe="")
    params = [("sdk", "1")]
    if return_url:
        params.append(("return_url", return_url))
    return base + "?" + urlencode(params)


def build_resource_redirect_url(
    video_ad_url: str,
    creator_hash: str,
    project_hash: str,
    resource_hash: str,
    return_url: Optional[str] = None,
) -> str:
    base = (
        _trim_slash(video_ad_url)
        + "/c/" + quote(creator_hash, safe="")
        + "/" + quote(project_hash, safe="")
        + "/" + quote(resource_hash, safe="")
    )
    params = [("sdk", "1")]
    if return_url:
        params.append(("return_url", return_url))
    return base + "?" + urlencode(params)


def remove_query_param(url: str, param: str) -> str:
    parsed = urlparse(url)
    query = [(k, v) for k, v in parse_qsl(parsed.query, keep_blank_values=True) if k != param]
    new_query = urlencode(query)
    return urlunparse(parsed._replace(query=new_query))


__all__ = [
    "build_video_ad_redirect_url",
    "build_resource_redirect_url",
    "remove_query_param",
]
